enhancements/etcd/storage-migration-for-etcd-encryption.md

```diff
@@ -0,0 +1,182 @@
+---
+title: storage-migration-for-etcd-encryption
+authors:
+  - "@sanchezl"
+reviewers:
+  - "@sttts"
+  - "@deads2k"
+  - "@enj"
+approvers:
+  - TBD
+creation-date: 2019-09-11
+last-updated: 2019-09-11
+status: provisional
+---
+
+# storage-migration-for-etcd-encryption
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift/docs]
+
+## Summary
+
+Introduce the `openshift-kube-storage-version-migrator-operator` to install the `StorageVersionMigration` CRD and the corresponding  `kube-storage-version-migrator` controller.  `StorageVersionMigration` resources can then be created by components, (initially  `kube-apiserver` and `openshift-apiserver`) to kick off storage migrations for resources.
+
+The first known usage is for etcd encryption at rest: `secrets` (`kube-apiserver`) and `oauthaccesstokens` (`openshift-apiserver`).
+
+## Motivation
+
+The immediate need for storage migration is to support the enablement of etcd data storage encryption. Encrypting etcd data storage involves rotating encryption keys periodically, which requires that stored data be migrated after choosing a new write key and before removing an old read key so that api servers can discard old encryption keys periodically and not lose the ability to decode the pre-existing data.
+
+We will use the upstream `kube-storage-version-migrator` (the controller-based successor to `oc adm migrate`). While upstream usage of the `kube-storage-version-migrator` is focused on migrating to new storage versions, we should be able to explicitly trigger the `kube-storage-version-migrator` when we need to migrate the encryption key used for storage, even when the storage version itself has not changed.
+
+## Goals
+
+* Make the `StorageVersionMigration` API and its corresponding controller available.
+* Invoke the `StorageVersionMigration` API from operators requesting that storage be migrated to use new etcd storage encrpyion keys.
+
+## Non-Goals
+
+* Will not provide a higher level API to the `StorageVersionMigration` api controlled by the `kube-storage-version-migrator`.
+* Will not implement any general control loops to automatically detect and then initiate a storage migration when needed.
+
+## Proposal
+
+* Introduce `openshift/cluster-openshift-kube-storage-version-migrator-operator` repository.
+  * Not required for bootstrapping or re-bootstrapping.
+  * No need for static pods.
+  * Will run on cluster.
+* Implement `openshift-kube-storage-version-migrator-operator`. 
+* Build `kube-storage-version-migrator` image.
+  * Fork `kubernetes-sigs/kube-storage-version-migrator` to `openshift`
```

Comment:
why fork? Just because of CI?

Reply:
> why fork? Just because of CI?

Yes.  To add dockerfiles and release information.  coredns did this previously.